[JSC] Rewrite Array#sort

Constellation opened the pull request:
#### 7726722f8ddd89a6c25ab071961cb059eab63d14
<pre>
[JSC] Rewrite Array#sort
<a href="https://bugs.webkit.org/show_bug.cgi?id=273051">https://bugs.webkit.org/show_bug.cgi?id=273051</a>
<a href="https://rdar.apple.com/126833827">rdar://126833827</a>

Reviewed by Keith Miller.

The world found that JS-implemented Array#sort is too slow, in particular in Baseline JIT.
As a result, all other engines rewrote Array#sort in C++ (V8 and SpiderMonkey). We also found that our Array#sort is too slow in Speedometer 3.0,
so we need to rewrite to follow to the other engines.

But now, the implementation is significantly simpler than before since Array#sort is now well defined in the spec and we first copy content to
non user-observable array for sorting. This avoids any changes to this array via function calls, so we can make sorting much simpler and
safer (even than JS implemented one).

1. VMTraps is moved to make access close to the other fields in VM.
2. Unify TypedArray / Array sort.
3. Add insertion sort path for smaller arrays.
4. Add a fast path skipping many comparisons by checking whether two subarrays are already sorted. This is aligned to V8 and SpiderMonkey behavior.

                                                         ToT                     Patched

    uint32array-sort-custom                        73.4050+-0.6354     ^     48.5713+-0.3494        ^ definitely 1.5113x faster
    array-prototype-sort-large-array               25.9099+-0.0528     ^     22.4718+-0.1368        ^ definitely 1.1530x faster
    array-prototype-sort-small-array-comparator
                                                   12.6596+-0.2183     ^      9.7617+-0.2233        ^ definitely 1.2969x faster
    array-prototype-sort-large-array-comparator
                                                   13.8265+-0.1290     ^      6.7050+-0.2796        ^ definitely 2.0621x faster
    sorting-benchmark                              10.0377+-0.1360     ^      8.3962+-0.2557        ^ definitely 1.1955x faster
    array-prototype-sort-medium-array-comparator
                                                   42.4683+-0.4189     ^     28.7110+-0.2369        ^ definitely 1.4792x faster
    array-prototype-sort-medium-array              70.2655+-0.6955     ^     29.8648+-0.3713        ^ definitely 2.3528x faster
    array-prototype-sort-small-array               20.3562+-0.1621     ^      9.7814+-0.0692        ^ definitely 2.0811x faster

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(linkTimeConstant.sortStringComparator): Deleted.
(linkTimeConstant.sortCompact): Deleted.
(linkTimeConstant.sortCommit): Deleted.
(linkTimeConstant.sortMerge): Deleted.
(linkTimeConstant.sortMergeSort): Deleted.
(linkTimeConstant.sortBucketSort): Deleted.
(sort): Deleted.
* Source/JavaScriptCore/heap/GCMemoryOperations.h:
(JSC::gcSafeMemcpy):
(JSC::gcSafeMemmove):
* Source/JavaScriptCore/interpreter/CachedCall.h:
(JSC::CachedCall::callWithArguments):
* Source/JavaScriptCore/runtime/ArgList.cpp:
(JSC::ArgList::getSlice const):
(JSC::MarkedVectorBase::markLists):
(JSC::MarkedVectorBase::expandCapacity):
* Source/JavaScriptCore/runtime/ArgList.h:
(JSC::MarkedVector::at const):
(JSC::MarkedVector::set):
(JSC::ArgList::ArgList):
(JSC::ArgList::at const):
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::sortCompact):
(JSC::sortBucketSort):
(JSC::sortStableSort):
(JSC::sortCommit):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::appendMemcpy):
* Source/JavaScriptCore/runtime/JSArray.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
(JSC::genericTypedArrayViewProtoFuncSortImpl):
(JSC::typedArrayMerge): Deleted.
(JSC::typedArrayMergeSort): Deleted.
(JSC::coerceComparatorResultToBoolean): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.cpp:
(JSC::JSTypedArrayViewPrototype::finishCreation):
* Source/JavaScriptCore/runtime/StableSort.h: Added.
(JSC::coerceComparatorResultToBoolean):
(JSC::arrayInsertionSort):
(JSC::arrayMerge):
(JSC::arrayStableSort):
* Source/JavaScriptCore/runtime/VM.h:

Canonical link: <a href="https://commits.webkit.org/277878@main">https://commits.webkit.org/277878@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae0c0a6b65f6dc6f66e622c505302320df36675d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/48846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28059 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51813 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51533 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44912 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34007 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25587 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/39934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/49428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/25710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/21036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/23173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/43298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6901 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/42145 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/45104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43792 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53444 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/48337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/23897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20166 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/25160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/46191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25968 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/55832 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6968 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24880 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/11494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (commented):
Nice

kmiller68 commented on Source/JavaScriptCore/runtime/StableSort.h:
```diff
@@ -0,0 +1,139 @@
+    auto* from = src.data();
+    size_t length = src.size();
+
+    constexpr size_t initialWidth = 4;
```
Is it worth making this configurable as a template parameter?

Constellation replied:
Sounds good. Changed.

kmiller68 commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -939,6 +944,271 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncSlice, (JSGlobalObject* globalObject, Cal
+        for (EncodedJSValue encodedValue : sorted) {
+            JSValue value = JSValue::decode(encodedValue);
+            constexpr bool shouldThrow = true;
+            thisObject->putByIndexInline(globalObject, index++, value, shouldThrow);
```
Does this hoist the array mode check out of the loop? If not maybe we should add such a function to do a bulk store. Maybe this is rare enough it doesn't matter?

Constellation replied:
Actually this code is rarely used. It is used only when appendMemcpy failed. And appendMemcpy already covers most of indexing type cases.

Constellation replied:
Also, since this can invoke user-defined setters, indexing mode can be changed during this loop.

kmiller68 commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -939,6 +944,271 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncSlice, (JSGlobalObject* globalObject, Cal
+    uint64_t undefinedMax = static_cast<uint64_t>(sorted.size()) + undefinedCount;
+    for (; index64 < undefinedMax; ++index64) {
+        constexpr bool shouldThrow = true;
+        thisObject->putByIndexInline(globalObject, index64, jsUndefined(), shouldThrow);
```
Ditto.

Constellation replied:
Ditto, this is rare code.

kmiller68 commented on Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
```diff
@@ -939,6 +944,271 @@ JSC_DEFINE_HOST_FUNCTION(arrayProtoFuncSlice, (JSGlobalObject* globalObject, Cal
+    }
+
+    for (; index64 < length; ++index64) {
+        bool deleted = thisObject->deleteProperty(globalObject, index64);
```
Ditto.

Constellation replied:
Ditto, this is rare code.

kmiller68 commented on Source/JavaScriptCore/runtime/JSArray.cpp:
```diff
@@ -479,6 +479,73 @@ bool JSArray::setLengthWithArrayStorage(JSGlobalObject* globalObject, unsigned n
     return true;
 }
 
+bool JSArray::appendMemcpy(JSGlobalObject* globalObject, VM& vm, unsigned startIndex, IndexingType otherType, std::span<const EncodedJSValue> values)
```
Nit: It feels like a lot of this could be deduplicated with the code below.

Constellation replied:
In fact, a lot of things are slightly different, and that's why they are separate. For example, double array format is different in these appendMemcpy functions.

kmiller68 commented on LayoutTests/js/dom/array-prototype-properties-expected.txt:
```diff
@@ -12,7 +12,7 @@ PASS Array.prototype.push.call(undefined, {}) threw exception TypeError: undefin
 PASS Array.prototype.shift.call(undefined) threw exception TypeError: undefined is not an object (evaluating 'Array.prototype.shift.call(undefined)').
 PASS Array.prototype.slice.call(undefined, 0, 1) threw exception TypeError: undefined is not an object (evaluating 'Array.prototype.slice.call(undefined, 0, 1)').
-PASS Array.prototype.sort.call(undefined) threw exception TypeError: Array.prototype.sort requires that |this| not be null or undefined.
+PASS Array.prototype.sort.call(undefined) threw exception TypeError: undefined is not an object (evaluating 'Array.prototype.sort.call(undefined)').
```
This feels like a regression.

Constellation replied:
I think newer is more consistent to the rest of the Array functions. So I'll leave it as is for now.

kmiller68 reviewed (approved):
r=me.

Constellation committed:
[JSC] Rewrite Array#sort
https://bugs.webkit.org/show_bug.cgi?id=273051
rdar://126833827

Reviewed by Keith Miller.

The world found that JS-implemented Array#sort is too slow, in particular in Baseline JIT.
As a result, all other engines rewrote Array#sort in C++ (V8 and SpiderMonkey). We also found that our Array#sort is too slow in Speedometer 3.0,
so we need to rewrite to follow to the other engines.

But now, the implementation is significantly simpler than before since Array#sort is now well defined in the spec and we first copy content to
non user-observable array for sorting. This avoids any changes to this array via function calls, so we can make sorting much simpler and
safer (even than JS implemented one).

1. VMTraps is moved to make access close to the other fields in VM.
2. Unify TypedArray / Array sort.
3. Add insertion sort path for smaller arrays.
4. Add a fast path skipping many comparisons by checking whether two subarrays are already sorted. This is aligned to V8 and SpiderMonkey behavior.

                                                         ToT                     Patched

    uint32array-sort-custom                        73.4050+-0.6354     ^     48.5713+-0.3494        ^ definitely 1.5113x faster
    array-prototype-sort-large-array               25.9099+-0.0528     ^     22.4718+-0.1368        ^ definitely 1.1530x faster
    array-prototype-sort-small-array-comparator
                                                   12.6596+-0.2183     ^      9.7617+-0.2233        ^ definitely 1.2969x faster
    array-prototype-sort-large-array-comparator
                                                   13.8265+-0.1290     ^      6.7050+-0.2796        ^ definitely 2.0621x faster
    sorting-benchmark                              10.0377+-0.1360     ^      8.3962+-0.2557        ^ definitely 1.1955x faster
    array-prototype-sort-medium-array-comparator
                                                   42.4683+-0.4189     ^     28.7110+-0.2369        ^ definitely 1.4792x faster
    array-prototype-sort-medium-array              70.2655+-0.6955     ^     29.8648+-0.3713        ^ definitely 2.3528x faster
    array-prototype-sort-small-array               20.3562+-0.1621     ^      9.7814+-0.0692        ^ definitely 2.0811x faster

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/builtins/ArrayPrototype.js:
(linkTimeConstant.sortStringComparator): Deleted.
(linkTimeConstant.sortCompact): Deleted.
(linkTimeConstant.sortCommit): Deleted.
(linkTimeConstant.sortMerge): Deleted.
(linkTimeConstant.sortMergeSort): Deleted.
(linkTimeConstant.sortBucketSort): Deleted.
(sort): Deleted.
* Source/JavaScriptCore/heap/GCMemoryOperations.h:
(JSC::gcSafeMemcpy):
(JSC::gcSafeMemmove):
* Source/JavaScriptCore/interpreter/CachedCall.h:
(JSC::CachedCall::callWithArguments):
* Source/JavaScriptCore/runtime/ArgList.cpp:
(JSC::ArgList::getSlice const):
(JSC::MarkedVectorBase::markLists):
(JSC::MarkedVectorBase::expandCapacity):
* Source/JavaScriptCore/runtime/ArgList.h:
(JSC::MarkedVector::at const):
(JSC::MarkedVector::set):
(JSC::ArgList::ArgList):
(JSC::ArgList::at const):
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::ArrayPrototype::finishCreation):
(JSC::sortCompact):
(JSC::sortBucketSort):
(JSC::sortStableSort):
(JSC::sortCommit):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/JSArray.cpp:
(JSC::JSArray::appendMemcpy):
* Source/JavaScriptCore/runtime/JSArray.h:
* Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h:
(JSC::genericTypedArrayViewProtoFuncSortImpl):
(JSC::typedArrayMerge): Deleted.
(JSC::typedArrayMergeSort): Deleted.
(JSC::coerceComparatorResultToBoolean): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSTypedArrayViewPrototype.cpp:
(JSC::JSTypedArrayViewPrototype::finishCreation):
* Source/JavaScriptCore/runtime/StableSort.h: Added.
(JSC::coerceComparatorResultToBoolean):
(JSC::arrayInsertionSort):
(JSC::arrayMerge):
(JSC::arrayStableSort):
* Source/JavaScriptCore/runtime/VM.h:

Canonical link: https://commits.webkit.org/277878@main
